src/ecmult_impl.h

```diff
@@ -403,4 +404,280 @@ static void secp256k1_ecmult(const secp256k1_ecmult_context *ctx, secp256k1_gej
+        node = parent;
+    }
+
+    heap->tree[node] = index;
+}
+
+static void secp256k1_sift_floyd(secp256k1_scalar_heap *heap, size_t node, uint32_t index) {
+    uint32_t child_index, other_index;
+    size_t child, other, half_size = heap->size >> 1;
+    const secp256k1_scalar *sc = heap->scalars;
+
+    while (node < half_size) {
+        /* Initially assume the left child is the larger child */
+        child = (node << 1) + 1;
+        child_index = heap->tree[child];
+
+        /* If there is a right child, check whether it's larger than the left */
+        other = child + 1;
+        if (other < heap->size) {
+            other_index = heap->tree[other];
+            if (secp256k1_scalar_cmp_var(&sc[other_index], &sc[child_index]) > 0) {
+                child = other;
+                child_index = other_index;
+            }
+        }
+
+        /* Move the larger child up, and recurse from its previous position */
+        heap->tree[node] = child_index;
+        node = child;
+    }
+
+    secp256k1_sift_up(heap, node, index);
+}
+
+SECP256K1_INLINE static void secp256k1_heapify(secp256k1_scalar_heap *heap) {
+    size_t root = heap->size >> 1;;
+    while (root-- > 0) {
+        secp256k1_sift_down(heap, root, heap->tree[root]);
+    }
+}
+
+static void secp256k1_heap_initialize(secp256k1_scalar_heap *heap, uint32_t *tree, const secp256k1_scalar *scalars, const secp256k1_gej *pt, size_t n) {
+    size_t i, size = 0;
+
+    heap->tree = tree;
+    for (i = 0; i < n; ++i) {
+        if (!secp256k1_scalar_is_zero(&scalars[i]) && !secp256k1_gej_is_infinity(&pt[i])) {
+            heap->tree[size++] = i;
+        }
+    }
+
+    heap->scalars = scalars;
+    heap->size = size;
+
+    secp256k1_heapify(heap);
+}
+
+SECP256K1_INLINE static uint32_t secp256k1_replace(secp256k1_scalar_heap *heap, uint32_t index) {
+    uint32_t result = heap->tree[0];
+    VERIFY_CHECK(heap->size > 0);
+    secp256k1_sift_floyd(heap, 0, index);
+    return result;
+}
+
+SECP256K1_INLINE static uint32_t secp256k1_heap_remove(secp256k1_scalar_heap *heap) {
+    uint32_t result = heap->tree[0];
+    VERIFY_CHECK(heap->size > 0);
+    if (--heap->size > 0) {
+        secp256k1_sift_down(heap, 0, heap->tree[heap->size]);
+    }
+    return result;
+}
+
+/** Multi-multiply: R = sum_i ni * Ai */
+static void secp256k1_ecmult_multi_bos_coster(uint32_t *tree_space, secp256k1_gej *r, secp256k1_scalar *sc, secp256k1_gej *pt, size_t n) {
+    secp256k1_scalar_heap heap;
+    uint32_t first, second;
+
+    secp256k1_gej_set_infinity(r);
+    secp256k1_heap_initialize(&heap, tree_space, sc, pt, n);
+
+    if (heap.size == 0) {
+        return;
+    }
+
+    first = secp256k1_heap_remove(&heap);
+
+    while (heap.size > 0) {
+        second = heap.tree[0];        
+
+        do {
+            /* Observe that nX + mY = (n-m)X + m(X + Y), and if n > m this transformation
+             * reduces the magnitude of the larger scalar, on average by half. So by
```

Comment:
On average by way more than half, if you have a sufficiently large number of points. If you have 256 randomly distributed 256-bit integers, I expect the difference between the top two to be around 2^253 or so only.